
# Actividad 4: Métodos Numéricos

## Universidad Externado de Colombia
### Ciencia de Datos

Esta actividad se centra en la aplicación de métodos numéricos para la resolución de ecuaciones no lineales. Se explorarán varios métodos como bisección, secante y Newton-Raphson a través de diferentes ejercicios.


In [10]:
import numpy as np
import math
import plotly.express as px
import time

### Carga de librerías

In [3]:
def biseccion(a, b, tol, max_iter, funcion):
    """
    Cálcula el método de la bisección dados dos puntos en una función.
    """
    if funcion(a) * funcion(b) > 0:
        print("La función no cambia de signo en el intervalo dado.")
        return None

    iteracion = 0
    raices = []
    while (b - a) / 2 > tol:
        midpoint = (a + b) / 2
        raices.append(midpoint)

        if funcion(midpoint) == 0:return midpoint, raices
        elif funcion(a) * funcion(midpoint) < 0: b = midpoint
        else: a = midpoint

        #print(f"Iteración {iteracion + 1}: Raíz aproximada = {midpoint}")
        iteracion += 1
        if iteracion >= max_iter:
            print("Se alcanzó el número máximo de iteraciones.")
            break

    return (a + b) / 2, ["Iteración " + str(i) + " " + str(raices[i]) for i in range(len(raices))]

In [4]:
def secante(x0, x1, tol, max_iter, funcion):
    """
    calcula el método de la secante para dos elemetnos de una función.
    """
    iter_count = 0
    while iter_count < max_iter:
        x_next = x1 - funcion(x1) * (x1 - x0) / (funcion(x1) - funcion(x0))
        print(abs(x_next - x1))
        if abs(x_next - x1) < tol:
            return x_next
        x0, x1 = x1, x_next
        iter_count += 1
    print("El método de la secante no converge.")
    return None


## Ejercicio 1

De acuerdo con el método de bisección, encuentre soluciones precisas dentro de $10^{-5}$ para las siguientes funciones:

- $x - 2^{-x} = 0$ para $[0, 1]$
- $e^x - x^2 + 3x - 2 = 0$ para $[0, 1]$
- $x \cos(x) - 2x^2 + 3x - 1 = 0$ para $[0.2, 0.3]$ y $[1.2, 1.3]$

Realice un esbozo de la gráfica junto con la iteración del método aplicado.


In [6]:
biseccion(1, 2, 0.00001, 1000, lambda x: 3*x - (math.e ** x))

(1.5121383666992188,
 ['Iteración 0 1.5',
  'Iteración 1 1.75',
  'Iteración 2 1.625',
  'Iteración 3 1.5625',
  'Iteración 4 1.53125',
  'Iteración 5 1.515625',
  'Iteración 6 1.5078125',
  'Iteración 7 1.51171875',
  'Iteración 8 1.513671875',
  'Iteración 9 1.5126953125',
  'Iteración 10 1.51220703125',
  'Iteración 11 1.511962890625',
  'Iteración 12 1.5120849609375',
  'Iteración 13 1.51214599609375',
  'Iteración 14 1.512115478515625',
  'Iteración 15 1.5121307373046875'])

In [7]:
biseccion(1, 2, 0.00001, 1000, lambda x: (2*x) + (3*math.cos(x)) - (math.e ** x))

(1.2397079467773438,
 ['Iteración 0 1.5',
  'Iteración 1 1.25',
  'Iteración 2 1.125',
  'Iteración 3 1.1875',
  'Iteración 4 1.21875',
  'Iteración 5 1.234375',
  'Iteración 6 1.2421875',
  'Iteración 7 1.23828125',
  'Iteración 8 1.240234375',
  'Iteración 9 1.2392578125',
  'Iteración 10 1.23974609375',
  'Iteración 11 1.239501953125',
  'Iteración 12 1.2396240234375',
  'Iteración 13 1.23968505859375',
  'Iteración 14 1.239715576171875',
  'Iteración 15 1.2397003173828125'])

In [29]:
biseccion(0, 1, 0.000001, 1000, lambda x:math.e**(x) - x**(2) + (3*x) - 2)

(0.25753116607666016,
 ['Iteración 0 0.5',
  'Iteración 1 0.25',
  'Iteración 2 0.375',
  'Iteración 3 0.3125',
  'Iteración 4 0.28125',
  'Iteración 5 0.265625',
  'Iteración 6 0.2578125',
  'Iteración 7 0.25390625',
  'Iteración 8 0.255859375',
  'Iteración 9 0.2568359375',
  'Iteración 10 0.25732421875',
  'Iteración 11 0.257568359375',
  'Iteración 12 0.2574462890625',
  'Iteración 13 0.25750732421875',
  'Iteración 14 0.257537841796875',
  'Iteración 15 0.2575225830078125',
  'Iteración 16 0.25753021240234375',
  'Iteración 17 0.2575340270996094',
  'Iteración 18 0.25753211975097656'])

In [13]:
def tiempoGrafica(x, y, f):
    print(time.time())

In [18]:
tiempoGrafica(5, 10, 3)

1709667300.4995022


In [ ]:
1-e5

In [30]:
biseccion(0.2, 0.3, 0.000001, 1000, lambda x:x*math.cos(x) - 2*(x**(2)) + (3*x) - 1)

(0.29753036499023444,
 ['Iteración 0 0.25',
  'Iteración 1 0.275',
  'Iteración 2 0.2875',
  'Iteración 3 0.29374999999999996',
  'Iteración 4 0.296875',
  'Iteración 5 0.2984375',
  'Iteración 6 0.29765625',
  'Iteración 7 0.297265625',
  'Iteración 8 0.2974609375',
  'Iteración 9 0.29755859375000004',
  'Iteración 10 0.29750976562500003',
  'Iteración 11 0.29753417968750007',
  'Iteración 12 0.29752197265625',
  'Iteración 13 0.29752807617187504',
  'Iteración 14 0.29753112792968756',
  'Iteración 15 0.29752960205078127'])

In [31]:
biseccion(1.2, 1.3, 0.000001, 1000, lambda x:x*math.cos(x) - 2*(x**(2)) + (3*x) - 1)

(1.2566230773925782,
 ['Iteración 0 1.25',
  'Iteración 1 1.275',
  'Iteración 2 1.2625',
  'Iteración 3 1.25625',
  'Iteración 4 1.259375',
  'Iteración 5 1.2578125',
  'Iteración 6 1.25703125',
  'Iteración 7 1.2566406250000002',
  'Iteración 8 1.2564453125000001',
  'Iteración 9 1.2565429687500003',
  'Iteración 10 1.2565917968750002',
  'Iteración 11 1.2566162109375',
  'Iteración 12 1.2566284179687501',
  'Iteración 13 1.256622314453125',
  'Iteración 14 1.2566253662109377',
  'Iteración 15 1.2566238403320313'])


## Ejercicio 2

Sea $f(x) = -x^3 - \cos(x)$. Con $p_0 = -1$ y $p_1 = 0$, encuentre $P_4$ utilizando el método de la secante.


In [35]:
secante(-1, 0, 1e-1, 4, lambda x:-x**(3) - math.cos(x))

0.6850733573260451
0.5670031315831837
0.44487095040313607
0.04057823092647639


-0.8477837694325692

In [43]:
%%html
<script src="https://cdn.geogebra.org/apps/deployggb.js"></script>

%%html

<div id="ggb-point"></div>

<script>
  var ggbApp = new GGBApplet({
      "height": 400,
      "showToolBar": false,
      "showMenuBar": false,
      "showAlgebraInput": false,
      "showResetIcon": true,
      "enableLabelDrags": false,
      "enableRightClick": false,
      "enableShiftDragZoom": true,
      "useBrowserForJS": false,
  }, 'ggb-point');

  ggbApp.inject();
</script>

In [45]:
def run_geogebra_command(command):
    display(Javascript(f"""
    ggbApplet.evalCommand("{command}");
    """))

# Trazar la función f(x) = x^2
run_geogebra_command("f(x) = x^2")


<IPython.core.display.Javascript object>


## Ejercicio 3

Realice el gráfico para $y = e^x - 2$ y $y = \cos(e^x - 2)$ y use el método de bisección para encontrar una aproximación dentro de $10^{-5}$ para $e^x - 2 = \cos(e^x - 2)$ en $[0.5, 1.5]$.



## Ejercicio 4

El polinomio de cuarto grado $f(x) = 230x^4 + 18x^3 + 9x^2 - 221x - 9$ tiene dos raíces reales, una en $[-1, 0]$ y otra en $[0, 1]$. Aproxime estos ceros con el método de la secante y de Newton.



## Ejercicio 5

Existen dos soluciones positivas $x_1$ y $x_2$ para la ecuación $4x^2 - e^x - e^{-x} = 0$. A partir del método de Newton, aproxime la solución dentro de 0.00001 de acuerdo con los siguientes valores de $p_0$:

- $p_0 = -10$
- $p_0 = 3$
- $p_0 = 10$
- $p_0 = -3$



## Ejercicio 6

La ecuación de anualidad ordinaria está dada por:

\[A = 
rac{P[1 - (1 + i)^{-n}]}{i}\]

Donde \(A\) es la cantidad de la hipoteca, \(P\) es la cantidad de cada pago, \(i\) es la tasa de interés por periodo para los \(n\) periodos de pago. Suponiendo una hipoteca por 540.000.000 COP a 30 años para una vivienda en la Candelaria y que la persona que solicita el préstamo puede realizar pagos de máximo 4.000.000 por mes, ¿cuál es la tasa de interés máxima que la persona puede asumir?
